server/region_syncer/server.go

```diff
@@ -0,0 +1,133 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/pd/server/core"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	msgSize                 = 8 * 1024 * 1024
+	maxSyncRegionBatchSize  = 100
+	syncerKeepAliveInterval = 10 * time.Second
+)
+
+// ClientStream is the client side of the region sycner.
+type ClientStream interface {
+	Recv() (*pdpb.SyncRegionResponse, error)
+	CloseSend() error
+}
+
+// ServerStream is the server side of the region syncer.
+type ServerStream interface {
+	Send(regions *pdpb.SyncRegionResponse) error
+}
+
+// Carrier is the abstraction of the syncer storage carrier.
+type Carrier interface {
+	Context() context.Context
+	ClusterID() uint64
+	GetMemberInfo() *pdpb.Member
+	GetLeader() *pdpb.Member
+	GetStorage() *core.KV
+}
+
+// RegionSyncer is used to sync the region information without raft.
+type RegionSyncer struct {
+	sync.RWMutex
+	streams map[string]ServerStream
+	ctx     context.Context
+	cancel  context.CancelFunc
+	carrier Carrier
+	closed  chan struct{}
+	wg      sync.WaitGroup
+}
+
+// NewRegionSyncer returns a region syncer.
+// The final consistency is ensured by the heartbeat.
+// Strong consistency is not guaranteed.
+// Usually open the region syncer in huge cluster and the carrier
+// no longer etcd but go-leveldb.
+func NewRegionSyncer(c Carrier) *RegionSyncer {
+	return &RegionSyncer{
+		streams: make(map[string]ServerStream),
+		carrier: c,
+		closed:  make(chan struct{}),
+	}
+}
+
+// RunServer runs the server of the region syncer.
+// regionNitifier is used to get the changed regions.
+func (s *RegionSyncer) RunServer(regionNotifier <-chan *core.RegionInfo, quit chan struct{}) {
+	var requests []*metapb.Region
+	ticker := time.NewTicker(syncerKeepAliveInterval)
+	for {
+		select {
+		case <-quit:
+			log.Info("exit region syncer")
+			return
+		case first := <-regionNotifier:
+			requests = append(requests, first.GetMeta())
+			pending := len(regionNotifier)
+			for i := 0; i < pending && i < maxSyncRegionBatchSize; i++ {
+				region := <-regionNotifier
+				requests = append(requests, region.GetMeta())
+			}
+			regions := &pdpb.SyncRegionResponse{
+				Header:  &pdpb.ResponseHeader{ClusterId: s.carrier.ClusterID()},
+				Regions: requests,
+			}
+			s.broadcast(regions)
+		case <-ticker.C:
+			alive := &pdpb.SyncRegionResponse{Header: &pdpb.ResponseHeader{ClusterId: s.carrier.ClusterID()}}
+			s.broadcast(alive)
+		}
+		requests = requests[:0]
+	}
+}
+
+// BindStream binds the established server stream.
+func (s *RegionSyncer) BindStream(name string, stream ServerStream) {
+	s.Lock()
+	defer s.Unlock()
+	s.streams[name] = stream
+}
+
+func (s *RegionSyncer) broadcast(regions *pdpb.SyncRegionResponse) {
+	failed := make([]string, 0, 3)
```

Comment:
Strange to hardcode the 3 here. I think `var failed []string` is better since it's unusual case fail when sending message.